packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -229,14 +235,18 @@ export class TokenDetectionController extends StaticIntervalPollingController<
     this.#disabled = disabled;
     this.setIntervalLength(interval);
 
-    this.#networkClientId = networkClientId;
     this.#selectedAddress =
       selectedAddress ??
       this.messagingSystem.call('AccountsController:getSelectedAccount')
         .address;
-    const { chainId } =
+    this.#addressAgainstWhichToDetect = this.#selectedAddress;
+
+    const { chainId, networkClientId: correctNetworkClientId } =
       this.#getCorrectChainIdAndNetworkClientId(networkClientId);
```

Comment:
With your changes to `#getCorrectChainIdAndNetworkClientId`, it will retrieve the currently selected network from NetworkController. So does the constructor need to take a `networkClientId`? It seems like we don't need to pass it here.

Reply:
Hmm looks like this is also related to https://github.com/MetaMask/core/pull/3938#discussion_r1509804249. If we allow `#getCorrectChainIdAndNetworkClientId` to accept a non-selected network client id and then return its chain id, then keeping the `networkClientId` constructor option would also allow token-detection-controller's internal state to diverge from network-controller's.

Removing this will be fairly disruptive for the tests, but it looks like it'll be necessary.

Reply:
Applied here: [`d0f6f69` (#3938)](https://github.com/MetaMask/core/pull/3938/commits/d0f6f69d4925ed0b71a16ba1b8fb78cd4238709d)